### PR TITLE
Ask for address and phone number on petitions#new 

### PR DESCRIPTION
**Pull request by ayghor**

### CODE REVIEW

File: spec/models/contact_spec.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  describe "associations" do
+    it { should belong_to(:signature) }
+  end
+
+  describe "validations" do
+    it { should validate_presence_of(:signature) }
+    it { should validate_presence_of(:address) }
+    it { should validate_presence_of(:phone_number) }
+    it { should validate_length_of(:address).is_at_most(500) }
+    it { should_not allow_value('6'*32).for(:phone_number).with_message(:too_long) }
```
Comment (pixeltrix) [inline spec/models/contact_spec.rb]: Again, expect > should - also add some air around the `'6'*32`, i.e. `'6' * 32` instead.

File: spec/models/contact_spec.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  end
+
+  describe "instance methods" do
+    describe "#phone_number=" do
+      it "normalizes input" do
+        contact = Contact.new(phone_number: "0300 200 6565")
+        expect(contact.phone_number).to eq("03002006565")
+      end
```
Comment (pixeltrix) [inline spec/models/contact_spec.rb]: That's implicitly relying on Active Record calling `#phone_number=` - better to use the method explicitly:

``` ruby
describe "#phone_number=" do
  it "normalizes input" do
    expect {
      subject.phone_number = "0300 200 6565"
    }.to change {
      subject.phone_number
    }.from(nil).to("03002006565")
  end
end
```

File: db/schema.rb [hunk 3]
```diff
@@ -214,6 +214,15 @@
     t.index ["petition_id", "constituency_id"], name: "idx_constituency_petition_journal_uniqueness", unique: true
   end
 
+  create_table "contacts", force: :cascade do |t|
+    t.bigint "signature_id", null: false
```
Comment (pixeltrix) [inline db/schema.rb]: Hmm, just noticed that some stuff is integer and some stuff is bigint - we should migrate to the latter but we can do that in another PR.

File: app/models/contact.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+  validates_length_of :address, maximum: 500
+  validates_length_of :phone_number, maximum: 31
+
+  def address=(x)
+    super(x.to_s.strip)
+  end
+
+  def phone_number=(x)
+    super(x.to_s.gsub(/[^0-9]/, ''))
```
[the remark this answers is not in the document]
Comment (ayghor) [reply app/models/contact.rb]: oh I didn't know you could use `^` with tr!

File: app/models/contact.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
+class Contact < ActiveRecord::Base
+  belongs_to :signature
+
+  validates_presence_of :signature
+  validates_presence_of :address
+  validates_presence_of :phone_number
+
+  validates_length_of :address, maximum: 500
+  validates_length_of :phone_number, maximum: 31
```
[the remark this answers is not in the document]
Comment (ayghor) [reply app/models/contact.rb]: Sure, I'll can change the style.

I separated the validations into their own lines because I thought it could make reviews easier. As if one alters the _presence_ validation, but not the _length_, then it would be obvious in the diff.

And in general, I've been feeling that they're simpler and map better to the Shoulda tests.

Thoughts?

[the remark this answers is not in the document]
Comment (ayghor) [reply app/models/contact.rb]: 31 chars would be 32 minus the trailing `'\0'`. It's just a _big enough_ length, that's also _not too big_. There's no DB restriction or any better reason for it.

File: features/charlie_creates_a_petition.feature [hunk 6]
```diff
@@ -127,13 +129,22 @@ Scenario: Charlie tries to submit an invalid petition
   Then I should see "Name must be completed"
   And I should see "Email must be completed"
   And I should see "Postcode must be completed"
+  And I should see "Phone number must be completed"
+  And I should see "Address must be completed"
+
+  When I fill in "Address" with "text longer than 500 characters aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
```
[the remark this answers is not in the document]
Comment (ayghor) [reply features/charlie_creates_a_petition.feature]: I thought it would be worse to spread this to another file, but yes, we can!

File: app/models/contact.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (ayghor) [reply app/models/contact.rb]: Also the specs fail if I remove the presence validation on **signature**:

```
  1) Contact validations should validate that :signature cannot be empty/falsy
     Failure/Error: it { should validate_presence_of(:signature) }

       Expected Contact to validate that :signature cannot be empty/falsy, but
       this could not be proved.
         After setting :signature to ‹nil›, the matcher expected the Contact to
         be invalid, but it was valid instead.
     # ./spec/models/contact_spec.rb:9:in `block (3 levels) in <top (required)>'
     # ./spec/support/database_cleaner.rb:10:in `block (3 levels) in <top (required)>'
     # /home/rogi/.gem/ruby/2.5.7/gems/database_cleaner-1.6.2/lib/database_cleaner/generic/base.rb:16:in `cleaning'
     # /home/rogi/.gem/ruby/2.5.7/gems/database_cleaner-1.6.2/lib/database_cleaner/base.rb:98:in `cleaning'
     # /home/rogi/.gem/ruby/2.5.7/gems/database_cleaner-1.6.2/lib/database_cleaner/configuration.rb:86:in `block (2 levels)
in cleaning'
     # /home/rogi/.gem/ruby/2.5.7/gems/database_cleaner-1.6.2/lib/database_cleaner/configuration.rb:87:in `cleaning'
     # ./spec/support/database_cleaner.rb:9:in `block (2 levels) in <top (required)>'
```

This is what I get on the console:

```
+irb(main):005:0> Contact.new.tap {|x| x.save; pp x.errors.messages }
   (0.5ms)  BEGIN
   (0.8ms)  ROLLBACK
{:address=>["Address must be completed"],
 :phone_number=>["Phone number must be completed"]}
```

File: app/models/petition_creator.rb [hunk 7]
```diff
@@ -106,6 +110,14 @@ def email
     petition_creator_params[:email].to_s.strip
   end
 
+  def phone_number
+    petition_creator_params[:phone_number].to_s.gsub(/[^0-9]/, '')
```
[the remark this answers is not in the document]
Comment (ayghor) [reply app/models/petition_creator.rb]: :sparkles: :sparkles: :sparkles:

File: app/models/signature.rb [hunk 8]
```diff
@@ -31,6 +31,7 @@ class Signature < ActiveRecord::Base
 
   belongs_to :petition
   belongs_to :invalidation
+  has_one :contact
```
[the remark this answers is not in the document]
Comment (ayghor) [reply app/models/signature.rb]: :sparkles: :sparkles: :sparkles:

File: app/models/contact.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (ayghor) [reply app/models/contact.rb]: :sparkles: :sparkles: :sparkles:

File: app/models/contact.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+  def address=(x)
```
[the remark this answers is not in the document]
Comment (ayghor) [reply app/models/contact.rb]: :sparkles: :sparkles: :sparkles:

File: spec/models/contact_spec.rb [hunk 10]
```diff
@@ -0,0 +1,24 @@
+require 'rails_helper'
+
+RSpec.describe Contact, type: :model do
+  describe "associations" do
+    it { should belong_to(:signature) }
```
[the remark this answers is not in the document]
Comment (ayghor) [reply spec/models/contact_spec.rb]: :sparkles: :sparkles: :sparkles:

File: spec/models/contact_spec.rb (hunk 1 above)
Comment (ayghor) [reply spec/models/contact_spec.rb]: :sparkles: :sparkles: :sparkles:

File: spec/models/contact_spec.rb (hunk 2 above)
Comment (ayghor) [reply spec/models/contact_spec.rb]: I it test the setter explicitly, but tweaked the style to what I believe to be more readable. Tell me what you think and I'll change it if you want.

File: features/charlie_creates_a_petition.feature (hunk 6 above)
[the remark this answers is not in the document]
Comment (ayghor) [reply features/charlie_creates_a_petition.feature]: it's done!

I found that the code validates that the text on a petition's *Additional details* is shorter than *800* characters, but the counter on the interface says *500*. Is this intended?

[the remark this answers is not in the document]
Comment (pixeltrix) [reply features/charlie_creates_a_petition.feature]: Yes, to allow for the fact that when textarea elements are submitted they submit new lines as CF/LF which gets inserted as two characters into the database so allow for some extra based on that.

File: spec/models/contact_spec.rb (hunk 2 above)
Comment (pixeltrix) [reply spec/models/contact_spec.rb]: Not a fan - can you change to how I posted it (which matches similar specs elsewhere).

File: app/models/contact.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (pixeltrix) [reply app/models/contact.rb]: Ahh, that's because we're not doing `load_defaults "5.2"` in config/application.rb

[the remark this answers is not in the document]
Comment (pixeltrix) [reply app/models/contact.rb]: image

Might be a lot of work to add `load_defaults` 🤔 

Best to leave the validation in place for now I think. 😂

File: spec/models/contact_spec.rb (hunk 2 above)
Comment (ayghor) [reply spec/models/contact_spec.rb]: fair enough :sparkles:  :sparkles:  :sparkles:
